// permissionless/_esm/actions/smartAccount/prepareUserOperationRequest.js
import { estimateFeesPerGas } from "viem/actions";
import { AccountOrClientNotFoundError, parseAccount } from "../../utils/index.js";
import { getAction } from "../../utils/getAction.js";
import { estimateUserOperationGas } from "../bundler/estimateUserOperationGas.js";
export async function prepareUserOperationRequest(client, args, stateOverrides) {
    const { account: account_ = client.account, userOperation: partialUserOperation, sponsorUserOperation } = args;
    if (!account_)
        throw new AccountOrClientNotFoundError();
    const account = parseAccount(account_);
    const [sender, nonce, initCode, callData, gasEstimation] = await Promise.all([
        partialUserOperation.sender || account.address,
        partialUserOperation.nonce || account.getNonce(),
        partialUserOperation.initCode || account.getInitCode(),
        partialUserOperation.callData,
        !partialUserOperation.maxFeePerGas ||
            !partialUserOperation.maxPriorityFeePerGas
            ? estimateFeesPerGas(account.client)
            : undefined
    ]);
    let userOperation = {
        sender,
        nonce,
        initCode,
        callData,
        paymasterAndData: "0x",
        signature: partialUserOperation.signature || "0x",
        maxFeePerGas: partialUserOperation.maxFeePerGas ||
            gasEstimation?.maxFeePerGas ||
            0n,
        maxPriorityFeePerGas: partialUserOperation.maxPriorityFeePerGas ||
            gasEstimation?.maxPriorityFeePerGas ||
            0n,
        callGasLimit: partialUserOperation.callGasLimit || 0n,
        verificationGasLimit: partialUserOperation.verificationGasLimit || 0n,
        preVerificationGas: partialUserOperation.preVerificationGas || 0n
    };
    if (userOperation.signature === "0x") {
        userOperation.signature = await account.getDummySignature(userOperation);
    }
    if (sponsorUserOperation) {
        userOperation = await sponsorUserOperation({
            userOperation,
            entryPoint: account.entryPoint
        });
    }
    else if (!userOperation.callGasLimit ||
        !userOperation.verificationGasLimit ||
        !userOperation.preVerificationGas) {
        const gasParameters = await getAction(client, estimateUserOperationGas)({
            userOperation: {
                ...userOperation
            },
            entryPoint: account.entryPoint
        }, stateOverrides);
        userOperation.callGasLimit =
            userOperation.callGasLimit || gasParameters.callGasLimit;
        userOperation.verificationGasLimit =
            userOperation.verificationGasLimit ||
                gasParameters.verificationGasLimit;
        userOperation.preVerificationGas =
            userOperation.preVerificationGas || gasParameters.preVerificationGas;
    }
    return userOperation;
}
