// permissionless/_cjs/actions/public/getSenderAddress.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSenderAddress = exports.InvalidEntryPointError = void 0;
const viem_1 = require("viem");
const actions_1 = require("viem/actions");
const getAction_1 = require("../../utils/getAction.js");
class InvalidEntryPointError extends viem_1.BaseError {
    constructor({ cause, entryPoint } = {}) {
        super(`The entry point address (\`entryPoint\`${entryPoint ? ` = ${entryPoint}` : ""}) is not a valid entry point. getSenderAddress did not revert with a SenderAddressResult error.`, {
            cause
        });
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: "InvalidEntryPointError"
        });
    }
}
exports.InvalidEntryPointError = InvalidEntryPointError;
const getSenderAddress = async (client, args) => {
    const { initCode, entryPoint } = args;
    try {
        await (0, getAction_1.getAction)(client, actions_1.simulateContract)({
            address: entryPoint,
            abi: [
                {
                    inputs: [
                        {
                            internalType: "address",
                            name: "sender",
                            type: "address"
                        }
                    ],
                    name: "SenderAddressResult",
                    type: "error"
                },
                {
                    inputs: [
                        {
                            internalType: "bytes",
                            name: "initCode",
                            type: "bytes"
                        }
                    ],
                    name: "getSenderAddress",
                    outputs: [],
                    stateMutability: "nonpayable",
                    type: "function"
                }
            ],
            functionName: "getSenderAddress",
            args: [initCode]
        });
    }
    catch (e) {
        const err = e;
        if (err.cause.name === "ContractFunctionRevertedError") {
            const revertError = err.cause;
            const errorName = revertError.data?.errorName ?? "";
            if (errorName === "SenderAddressResult" &&
                revertError.data?.args &&
                revertError.data?.args[0]) {
                return revertError.data?.args[0];
            }
        }
        throw e;
    }
    throw new InvalidEntryPointError({ entryPoint });
};
exports.getSenderAddress = getSenderAddress;
