// permissionless/_cjs/actions/public/getAccountNonce.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccountNonce = void 0;
const actions_1 = require("viem/actions");
const getAction_1 = require("../../utils/getAction.js");
const getAccountNonce = async (client, args) => {
    const { sender, entryPoint, key = BigInt(0) } = args;
    return await (0, getAction_1.getAction)(client, actions_1.readContract)({
        address: entryPoint,
        abi: [
            {
                inputs: [
                    {
                        name: "sender",
                        type: "address"
                    },
                    {
                        name: "key",
                        type: "uint192"
                    }
                ],
                name: "getNonce",
                outputs: [
                    {
                        name: "nonce",
                        type: "uint256"
                    }
                ],
                stateMutability: "view",
                type: "function"
            }
        ],
        functionName: "getNonce",
        args: [sender, key]
    });
};
exports.getAccountNonce = getAccountNonce;
