// permissionless/_cjs/actions/smartAccount/prepareUserOperationRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepareUserOperationRequest = void 0;
const actions_1 = require("viem/actions");
const utils_1 = require("../../utils/index.js");
const getAction_1 = require("../../utils/getAction.js");
const estimateUserOperationGas_1 = require("../bundler/estimateUserOperationGas.js");
async function prepareUserOperationRequest(client, args, stateOverrides) {
    const { account: account_ = client.account, userOperation: partialUserOperation, sponsorUserOperation } = args;
    if (!account_)
        throw new utils_1.AccountOrClientNotFoundError();
    const account = (0, utils_1.parseAccount)(account_);
    const [sender, nonce, initCode, callData, gasEstimation] = await Promise.all([
        partialUserOperation.sender || account.address,
        partialUserOperation.nonce || account.getNonce(),
        partialUserOperation.initCode || account.getInitCode(),
        partialUserOperation.callData,
        !partialUserOperation.maxFeePerGas ||
            !partialUserOperation.maxPriorityFeePerGas
            ? (0, actions_1.estimateFeesPerGas)(account.client)
            : undefined
    ]);
    let userOperation = {
        sender,
        nonce,
        initCode,
        callData,
        paymasterAndData: "0x",
        signature: partialUserOperation.signature || "0x",
        maxFeePerGas: partialUserOperation.maxFeePerGas ||
            gasEstimation?.maxFeePerGas ||
            0n,
        maxPriorityFeePerGas: partialUserOperation.maxPriorityFeePerGas ||
            gasEstimation?.maxPriorityFeePerGas ||
            0n,
        callGasLimit: partialUserOperation.callGasLimit || 0n,
        verificationGasLimit: partialUserOperation.verificationGasLimit || 0n,
        preVerificationGas: partialUserOperation.preVerificationGas || 0n
    };
    if (userOperation.signature === "0x") {
        userOperation.signature = await account.getDummySignature(userOperation);
    }
    if (sponsorUserOperation) {
        userOperation = await sponsorUserOperation({
            userOperation,
            entryPoint: account.entryPoint
        });
    }
    else if (!userOperation.callGasLimit ||
        !userOperation.verificationGasLimit ||
        !userOperation.preVerificationGas) {
        const gasParameters = await (0, getAction_1.getAction)(client, estimateUserOperationGas_1.estimateUserOperationGas)({
            userOperation: {
                ...userOperation
            },
            entryPoint: account.entryPoint
        }, stateOverrides);
        userOperation.callGasLimit =
            userOperation.callGasLimit || gasParameters.callGasLimit;
        userOperation.verificationGasLimit =
            userOperation.verificationGasLimit ||
                gasParameters.verificationGasLimit;
        userOperation.preVerificationGas =
            userOperation.preVerificationGas || gasParameters.preVerificationGas;
    }
    return userOperation;
}
exports.prepareUserOperationRequest = prepareUserOperationRequest;
