// permissionless/_cjs/actions/pimlico/sponsorUserOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sponsorUserOperation = void 0;
const deepHexlify_1 = require("../../utils/deepHexlify.js");
const sponsorUserOperation = async (client, args) => {
    const response = await client.request({
        method: "pm_sponsorUserOperation",
        params: args.sponsorshipPolicyId
            ? [
                (0, deepHexlify_1.deepHexlify)(args.userOperation),
                args.entryPoint,
                {
                    sponsorshipPolicyId: args.sponsorshipPolicyId
                }
            ]
            : [
                (0, deepHexlify_1.deepHexlify)(args.userOperation),
                args.entryPoint
            ]
    });
    const userOperation = {
        ...args.userOperation,
        paymasterAndData: response.paymasterAndData,
        preVerificationGas: BigInt(response.preVerificationGas),
        verificationGasLimit: BigInt(response.verificationGasLimit),
        callGasLimit: BigInt(response.callGasLimit)
    };
    return userOperation;
};
exports.sponsorUserOperation = sponsorUserOperation;
