// permissionless/_cjs/accounts/simple/signerToSimpleSmartAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signerToSimpleSmartAccount = void 0;
const viem_1 = require("viem");
const accounts_1 = require("viem/accounts");
const actions_1 = require("viem/actions");
const getAccountNonce_1 = require("../../actions/public/getAccountNonce.js");
const getSenderAddress_1 = require("../../actions/public/getSenderAddress.js");
const getUserOperationHash_1 = require("../../utils/getUserOperationHash.js");
const isSmartAccountDeployed_1 = require("../../utils/isSmartAccountDeployed.js");
const types_1 = require("../types.js");
const getAccountInitCode = async (factoryAddress, owner, index = 0n) => {
    if (!owner)
        throw new Error("Owner account not found");
    return (0, viem_1.concatHex)([
        factoryAddress,
        (0, viem_1.encodeFunctionData)({
            abi: [
                {
                    inputs: [
                        {
                            internalType: "address",
                            name: "owner",
                            type: "address"
                        },
                        {
                            internalType: "uint256",
                            name: "salt",
                            type: "uint256"
                        }
                    ],
                    name: "createAccount",
                    outputs: [
                        {
                            internalType: "contract SimpleAccount",
                            name: "ret",
                            type: "address"
                        }
                    ],
                    stateMutability: "nonpayable",
                    type: "function"
                }
            ],
            functionName: "createAccount",
            args: [owner, index]
        })
    ]);
};
const getAccountAddress = async ({ client, factoryAddress, entryPoint, owner, index = 0n }) => {
    const initCode = await getAccountInitCode(factoryAddress, owner, index);
    return (0, getSenderAddress_1.getSenderAddress)(client, {
        initCode,
        entryPoint
    });
};
async function signerToSimpleSmartAccount(client, { signer, factoryAddress, entryPoint, index = 0n, address }) {
    const viemSigner = {
        ...signer,
        signTransaction: (_, __) => {
            throw new types_1.SignTransactionNotSupportedBySmartAccount();
        }
    };
    const [accountAddress, chainId] = await Promise.all([
        address ??
            getAccountAddress({
                client,
                factoryAddress,
                entryPoint,
                owner: viemSigner.address,
                index
            }),
        (0, actions_1.getChainId)(client)
    ]);
    if (!accountAddress)
        throw new Error("Account address not found");
    let smartAccountDeployed = await (0, isSmartAccountDeployed_1.isSmartAccountDeployed)(client, accountAddress);
    const account = (0, accounts_1.toAccount)({
        address: accountAddress,
        async signMessage({ message }) {
            return (0, actions_1.signMessage)(client, { account: viemSigner, message });
        },
        async signTransaction(_, __) {
            throw new types_1.SignTransactionNotSupportedBySmartAccount();
        },
        async signTypedData(typedData) {
            return (0, actions_1.signTypedData)(client, {
                account: viemSigner,
                ...typedData
            });
        }
    });
    return {
        ...account,
        client: client,
        publicKey: accountAddress,
        entryPoint: entryPoint,
        source: "SimpleSmartAccount",
        async getNonce() {
            return (0, getAccountNonce_1.getAccountNonce)(client, {
                sender: accountAddress,
                entryPoint: entryPoint
            });
        },
        async signUserOperation(userOperation) {
            return account.signMessage({
                message: {
                    raw: (0, getUserOperationHash_1.getUserOperationHash)({
                        userOperation,
                        entryPoint: entryPoint,
                        chainId: chainId
                    })
                }
            });
        },
        async getInitCode() {
            if (smartAccountDeployed)
                return "0x";
            smartAccountDeployed = await (0, isSmartAccountDeployed_1.isSmartAccountDeployed)(client, accountAddress);
            if (smartAccountDeployed)
                return "0x";
            return getAccountInitCode(factoryAddress, viemSigner.address, index);
        },
        async encodeDeployCallData(_) {
            throw new Error("Simple account doesn't support account deployment");
        },
        async encodeCallData(args) {
            if (Array.isArray(args)) {
                const argsArray = args;
                return (0, viem_1.encodeFunctionData)({
                    abi: [
                        {
                            inputs: [
                                {
                                    internalType: "address[]",
                                    name: "dest",
                                    type: "address[]"
                                },
                                {
                                    internalType: "bytes[]",
                                    name: "func",
                                    type: "bytes[]"
                                }
                            ],
                            name: "executeBatch",
                            outputs: [],
                            stateMutability: "nonpayable",
                            type: "function"
                        }
                    ],
                    functionName: "executeBatch",
                    args: [
                        argsArray.map((a) => a.to),
                        argsArray.map((a) => a.data)
                    ]
                });
            }
            const { to, value, data } = args;
            return (0, viem_1.encodeFunctionData)({
                abi: [
                    {
                        inputs: [
                            {
                                internalType: "address",
                                name: "dest",
                                type: "address"
                            },
                            {
                                internalType: "uint256",
                                name: "value",
                                type: "uint256"
                            },
                            {
                                internalType: "bytes",
                                name: "func",
                                type: "bytes"
                            }
                        ],
                        name: "execute",
                        outputs: [],
                        stateMutability: "nonpayable",
                        type: "function"
                    }
                ],
                functionName: "execute",
                args: [to, value, data]
            });
        },
        async getDummySignature(_userOperation) {
            return "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
        }
    };
}
exports.signerToSimpleSmartAccount = signerToSimpleSmartAccount;
