// permissionless/_cjs/accounts/kernel/signerToEcdsaKernelSmartAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signerToEcdsaKernelSmartAccount = void 0;
const viem_1 = require("viem");
const actions_1 = require("viem/actions");
const getAccountNonce_1 = require("../../actions/public/getAccountNonce.js");
const getSenderAddress_1 = require("../../actions/public/getSenderAddress.js");
const utils_1 = require("../../utils/index.js");
const getUserOperationHash_1 = require("../../utils/getUserOperationHash.js");
const isSmartAccountDeployed_1 = require("../../utils/isSmartAccountDeployed.js");
const toSmartAccount_1 = require("../toSmartAccount.js");
const types_1 = require("../types.js");
const KernelAccountAbi_1 = require("./abi/KernelAccountAbi.js");
const createAccountAbi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "_implementation",
                type: "address"
            },
            {
                internalType: "bytes",
                name: "_data",
                type: "bytes"
            },
            {
                internalType: "uint256",
                name: "_index",
                type: "uint256"
            }
        ],
        name: "createAccount",
        outputs: [
            {
                internalType: "address",
                name: "proxy",
                type: "address"
            }
        ],
        stateMutability: "payable",
        type: "function"
    }
];
const KERNEL_ADDRESSES = {
    ECDSA_VALIDATOR: "0xd9AB5096a832b9ce79914329DAEE236f8Eea0390",
    ACCOUNT_V2_2_LOGIC: "0x0DA6a956B9488eD4dd761E59f52FDc6c8068E6B5",
    FACTORY_ADDRESS: "0x5de4839a76cf55d0c90e2061ef4386d962E15ae3"
};
const getAccountInitCode = async ({ owner, index, accountLogicAddress, ecdsaValidatorAddress }) => {
    if (!owner)
        throw new Error("Owner account not found");
    const initialisationData = (0, viem_1.encodeFunctionData)({
        abi: KernelAccountAbi_1.KernelInitAbi,
        functionName: "initialize",
        args: [ecdsaValidatorAddress, owner]
    });
    return (0, viem_1.encodeFunctionData)({
        abi: createAccountAbi,
        functionName: "createAccount",
        args: [accountLogicAddress, initialisationData, index]
    });
};
const getAccountAddress = async ({ client, owner, entryPoint: entryPointAddress, initCodeProvider, ecdsaValidatorAddress, deployedAccountAddress, factoryAddress }) => {
    if (deployedAccountAddress !== undefined) {
        const deployedAccountOwner = await (0, actions_1.readContract)(client, {
            address: ecdsaValidatorAddress,
            abi: [
                {
                    inputs: [
                        {
                            internalType: "address",
                            name: "",
                            type: "address"
                        }
                    ],
                    name: "ecdsaValidatorStorage",
                    outputs: [
                        {
                            internalType: "address",
                            name: "owner",
                            type: "address"
                        }
                    ],
                    stateMutability: "view",
                    type: "function"
                }
            ],
            functionName: "ecdsaValidatorStorage",
            args: [deployedAccountAddress]
        });
        if (!(0, viem_1.isAddressEqual)(deployedAccountOwner, owner)) {
            throw new Error("Invalid owner for the already deployed account");
        }
        return deployedAccountAddress;
    }
    const factoryData = await initCodeProvider();
    return (0, getSenderAddress_1.getSenderAddress)(client, {
        initCode: (0, viem_1.concatHex)([factoryAddress, factoryData]),
        entryPoint: entryPointAddress
    });
};
async function signerToEcdsaKernelSmartAccount(client, { signer, address, entryPoint: entryPointAddress, index = 0n, factoryAddress = KERNEL_ADDRESSES.FACTORY_ADDRESS, accountLogicAddress = KERNEL_ADDRESSES.ACCOUNT_V2_2_LOGIC, ecdsaValidatorAddress = KERNEL_ADDRESSES.ECDSA_VALIDATOR, deployedAccountAddress }) {
    const entryPointVersion = (0, utils_1.getEntryPointVersion)(entryPointAddress);
    if (entryPointVersion !== "v0.6") {
        throw new Error("Only EntryPoint 0.6 is supported");
    }
    const viemSigner = {
        ...signer,
        signTransaction: (_, __) => {
            throw new types_1.SignTransactionNotSupportedBySmartAccount();
        }
    };
    const generateInitCode = () => getAccountInitCode({
        owner: viemSigner.address,
        index,
        accountLogicAddress,
        ecdsaValidatorAddress
    });
    const [accountAddress, chainId] = await Promise.all([
        address ??
            getAccountAddress({
                client,
                entryPoint: entryPointAddress,
                owner: viemSigner.address,
                ecdsaValidatorAddress,
                initCodeProvider: generateInitCode,
                deployedAccountAddress,
                factoryAddress
            }),
        (0, actions_1.getChainId)(client)
    ]);
    if (!accountAddress)
        throw new Error("Account address not found");
    let smartAccountDeployed = await (0, isSmartAccountDeployed_1.isSmartAccountDeployed)(client, accountAddress);
    return (0, toSmartAccount_1.toSmartAccount)({
        address: accountAddress,
        async signMessage({ message }) {
            return (0, actions_1.signMessage)(client, { account: viemSigner, message });
        },
        async signTransaction(_, __) {
            throw new types_1.SignTransactionNotSupportedBySmartAccount();
        },
        async signTypedData(typedData) {
            return (0, actions_1.signTypedData)(client, {
                account: viemSigner,
                ...typedData
            });
        },
        client: client,
        publicKey: accountAddress,
        entryPoint: entryPointAddress,
        source: "kernelEcdsaSmartAccount",
        async getNonce() {
            return (0, getAccountNonce_1.getAccountNonce)(client, {
                sender: accountAddress,
                entryPoint: entryPointAddress
            });
        },
        async signUserOperation(userOperation) {
            const hash = (0, getUserOperationHash_1.getUserOperationHash)({
                userOperation: {
                    ...userOperation,
                    signature: "0x"
                },
                entryPoint: entryPointAddress,
                chainId: chainId
            });
            const signature = await (0, actions_1.signMessage)(client, {
                account: viemSigner,
                message: { raw: hash }
            });
            return (0, viem_1.concatHex)(["0x00000000", signature]);
        },
        async getInitCode() {
            if (smartAccountDeployed)
                return "0x";
            smartAccountDeployed = await (0, isSmartAccountDeployed_1.isSmartAccountDeployed)(client, accountAddress);
            if (smartAccountDeployed)
                return "0x";
            return (0, viem_1.concatHex)([factoryAddress, await generateInitCode()]);
        },
        async getFactory() {
            if (smartAccountDeployed)
                return undefined;
            smartAccountDeployed = await (0, isSmartAccountDeployed_1.isSmartAccountDeployed)(client, accountAddress);
            if (smartAccountDeployed)
                return undefined;
            return factoryAddress;
        },
        async getFactoryData() {
            if (smartAccountDeployed)
                return undefined;
            smartAccountDeployed = await (0, isSmartAccountDeployed_1.isSmartAccountDeployed)(client, accountAddress);
            if (smartAccountDeployed)
                return undefined;
            return generateInitCode();
        },
        async encodeDeployCallData(_) {
            throw new Error("Simple account doesn't support account deployment");
        },
        async encodeCallData(_tx) {
            if (Array.isArray(_tx)) {
                return (0, viem_1.encodeFunctionData)({
                    abi: KernelAccountAbi_1.KernelExecuteAbi,
                    functionName: "executeBatch",
                    args: [
                        _tx.map((tx) => ({
                            to: tx.to,
                            value: tx.value,
                            data: tx.data
                        }))
                    ]
                });
            }
            return (0, viem_1.encodeFunctionData)({
                abi: KernelAccountAbi_1.KernelExecuteAbi,
                functionName: "execute",
                args: [_tx.to, _tx.value, _tx.data, 0]
            });
        },
        async getDummySignature(_userOperation) {
            return "0x00000000fffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
        }
    });
}
exports.signerToEcdsaKernelSmartAccount = signerToEcdsaKernelSmartAccount;
